src/rhsmlib/services/entitlement.py

```diff
@@ -0,0 +1,128 @@
+            return {'status': overall_status, 'reasons': reasons, 'valid': valid}
+        else:
+            return {'status': 'Unknown', 'reasons': {}, 'valid': False}
+
+    def get_pools(self, pool_subsets=None, matches=None, pool_only=None, match_installed=None,
+            no_overlap=None, service_level=None, show_all=None, on_date=None, **kwargs):
+        # We accept a **kwargs argument so that the DBus object can pass whatever dictionary it receives
+        # via keyword expansion.
+        if kwargs:
+            raise exceptions.ValidationError(_("Unknown arguments: %s") % kwargs.keys())
+
+        # [] or None means look at all pools
+        if not pool_subsets:
+            pool_subsets = ['installed', 'consumed', 'available']
+
+        options = {
+            'pool_subsets': pool_subsets,
+            'matches': matches,
+            'pool_only': pool_only,
+            'match_installed': match_installed,
+            'no_overlap': no_overlap,
+            'service_level': service_level,
+            'show_all': show_all,
+            'on_date': on_date,
+        }
+        self.validate_options(options)
+        results = {}
+        if 'installed' in pool_subsets:
+            installed = products.InstalledProducts(self.cp).list(options['matches'])
+            results['installed'] = [x._asdict() for x in installed]
+        if 'consumed' in pool_subsets:
+            results['consumed'] = self.get_consumed_product_pools(**options)
+        if 'available' in pool_subsets:
+            results['available'] = self.get_available_pools(**options)
+
+        return results
+
+    def get_consumed_product_pools(self, service_level=None, matches=None, **kwargs):
+        certs = self.entitlement_dir.list()
+        cert_filter = utils.EntitlementCertificateFilter(filter_string=matches, service_level=service_level)
+
+        if service_level is not None or matches is not None:
+            certs = list(filter(cert_filter.match, certs))
+
+        return certs
+
+    def get_available_pools(self, show_all=None, on_date=None, no_overlap=None,
+            match_installed=None, matches=None, service_level=None, **kwargs):
+        available_pools = managerlib.get_available_entitlements(
+            get_all=show_all,
+            active_on=on_date,
+            overlapping=no_overlap,
+            uninstalled=match_installed,
+            filter_string=matches
+        )
+
+        def filter_pool_by_service_level(pool_data):
+            pool_level = ""
+            if pool_data['service_level']:
+                pool_level = pool_data['service_level']
+            return service_level.lower() == pool_level.lower()
+
+        if service_level is not None:
+            available_pools = list(filter(filter_pool_by_service_level, available_pools))
+
+        return available_pools
+
+    def validate_options(self, options):
+        if not set(['installed', 'consumed', 'available']).issuperset(options['pool_subsets']):
```

Comment:
When you run following command:  `$ dbus-send --system --print-reply  --dest='com.redhat.RHSM1' '/com/redhat/RHSM1/Entitlement' com.redhat.RHSM1.Entitlement.GetPools dict:string:string:"pool_subsets","installed" dict:string:string:"",""` then you will get following error:

```console
Error com.redhat.RHSM1.Error: Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/dbus/service.py", line 707, in _message_cb
    retval = candidate_method(self, *args, **keywords)
  File "<string>", line 2, in GetPools
  File "/home/jiri/github/candlepin/subscription_manager/src/rhsmlib/dbus/util.py", line 43, in dbus_handle_exceptions
    six.reraise(exceptions.RHSM1DBusException, "%s: %s" % (type(e).__name__, str(e)), trace)
  File "/home/jiri/github/candlepin/subscription_manager/src/rhsmlib/dbus/util.py", line 35, in dbus_handle_exceptions
    ret = func(*args, **kwargs)
  File "/home/jiri/github/candlepin/subscription_manager/src/rhsmlib/dbus/objects/entitlement.py", line 77, in GetPools
    pools = entitlement_service.get_pools(**options)
  File "/home/jiri/github/candlepin/subscription_manager/src/rhsmlib/services/entitlement.py", line 64, in get_pools
    self.validate_options(options)
  File "/home/jiri/github/candlepin/subscription_manager/src/rhsmlib/services/entitlement.py", line 109, in validate_options
    raise exceptions.ValidationError(_('Error: invalid listing type provided.  Only "installed", '
RHSM1DBusException: com.redhat.RHSM1.Error: ValidationError: Error: invalid listing type provided.  Only "installed", "consumed", or "available" are allowed
```

because `options['pool_subsets']` is string not set. If you change this line to this:

```python
options['pool_subsets'] = set([options['pool_subsets']])
if not set(['installed', 'consumed', 'available']).issuperset(options['pool_subsets']):
```

then it solve the problem, but consider the example of code only as a hint :-).